---
#### Nama: Aisyah Fadhillah Muhammad Azis

#### NIM: 241061013

#### Mata Kuliah: Teknik Pengambilan Sampel dan Data Wrangling

#### Tugas 2 Data Preparation
---

In [67]:
#Mengimport Library
import pandas as pd
import numpy as np

#Menampilkan dataset
df = pd.read_csv('data_tugas 2.csv')
df.head()

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
1,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
2,?,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.6,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.6,66.4,54.3


In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   normalized-losses  205 non-null    object 
 1   make               201 non-null    object 
 2   bahan-bakar        192 non-null    object 
 3   jumlah-pintu       189 non-null    object 
 4   body-style         193 non-null    object 
 5   wheel-bases        202 non-null    float64
 6   length             197 non-null    float64
 7   width              200 non-null    float64
 8   height             201 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.5+ KB


Dataset yang digunakan terdiri dari **205 baris** (observasi) dan **9 kolom** (variabel), yaitu normalized-losses, make, bahan-bakar, jumlah-pintu, body-style, wheel-bases, length, width, dan height. Berdasarkan tipe datanya, variabel dalam dataset ini terbagi menjadi dua jenis, yaitu **variabel kategorik (object)** yang meliputi normalized-losses, make, bahan-bakar, jumlah-pintu, dan body-style, serta **variabel numerik (float64)** yang terdiri dari wheel-bases, length, width, dan height. Hasil pemeriksaan awal juga menunjukkan bahwa terdapat beberapa data yang belum lengkap (missing value) pada sejumlah variabel, sehingga diperlukan proses data cleaning sebelum dilakukan analisis lebih lanjut.

##1. Menampilkan Jumlah Data Hilang
Lakukan pemeriksaan terhadap dataset untuk mengetahui jumlah **data hilang (missing values)** pada setiap variabel.

**Instruksi:**
*   Tampilkan jumlah missing value pada setiap kolom.
*   Jelaskan variabel mana yang memiliki data hilang paling banyak.

**Output yang diharapkan:**

Tabel jumlah missing value setiap variabel.


In [69]:
# deteksi pola missing non-standar
missing_patterns = ['?', 'NA', 'N/A', '-', '--', '', 'null', 'None']

# konversi ke NaN
df.replace(missing_patterns, np.nan, inplace=True)

# tabel missing value
missing_table = df.isnull().sum().to_frame(name='Jumlah Missing')
missing_table.sort_values(by='Jumlah Missing', ascending=False)

,Jumlah Missing
normalized-losses,41
jumlah-pintu,18
bahan-bakar,13
body-style,12
length,8
width,5
make,4
height,4
wheel-bases,3


Hasil pemeriksaan missing value menunjukkan bahwa variabel normalized-losses memiliki jumlah data hilang paling banyak, yaitu sebanyak 41 data, diikuti oleh jumlah-pintu, bahan-bakar, dan body-style, dst.

In [70]:
# Mengubah 'normalized-losses' ke numerik agar dapat diproses lebih lanjut,
# karena sebelumnya masih bertipe object akibat adanya nilai non-standar.

df['normalized-losses'] = pd.to_numeric(df['normalized-losses'], errors='coerce')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   normalized-losses  164 non-null    float64
 1   make               201 non-null    object 
 2   bahan-bakar        192 non-null    object 
 3   jumlah-pintu       187 non-null    object 
 4   body-style         193 non-null    object 
 5   wheel-bases        202 non-null    float64
 6   length             197 non-null    float64
 7   width              200 non-null    float64
 8   height             201 non-null    float64
dtypes: float64(5), object(4)
memory usage: 14.5+ KB


##2. Mengganti Missing Value pada Variabel Kategori
Untuk variabel berbentuk **kategori**, data hilang harus diganti dengan kategori baru yaitu **"unknown"**.

**Variabel kategorik pada dataset:**
*   make
*   bahan-bakar
*   jumlah-pintu
*   body-style

**Instruksi:**
*   Identifikasi missing value pada variabel kategori
*   Ganti nilai tersebut dengan **"unknown"**

**Output yang diharapkan:**

Tabel ringkasan jumlah setiap kategori pada keempat variabel


In [71]:
# Mengecek jumlah missing value pada variabel kategorik
kategori_cols = ['make', 'bahan-bakar', 'jumlah-pintu', 'body-style']

df[kategori_cols].isnull().sum()

,0
make,4
bahan-bakar,13
jumlah-pintu,18
body-style,12


In [72]:
# Mengganti missing value pada variabel kategorik dengan "unknown"
df[kategori_cols] = df[kategori_cols].fillna('unknown')

In [73]:
# Memastikan bahwa seluruh missing value pada variabel kategorik
# telah berhasil digantikan dengan "unknown"
df[kategori_cols].isnull().sum()

,0
make,0
bahan-bakar,0
jumlah-pintu,0
body-style,0


In [74]:
# Menampilkan distribusi setiap kategori untuk melihat perubahan data
# setelah proses penggantian missing value
for col in kategori_cols:
  print(f"\nDistribusi {col}:")
  print(df[col].value_counts())


Distribusi make:
make
toyota           32
mazda            17
nissan           17
honda            13
mitsubishi       12
volkswagen       12
volvo            11
subaru           11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           4
unknown           4
isuzu             4
alfa-romero       3
chevrolet         3
jaguar            3
renault           2
mercury           1
Name: count, dtype: int64

Distribusi bahan-bakar:
bahan-bakar
gas        173
diesel      19
unknown     13
Name: count, dtype: int64

Distribusi jumlah-pintu:
jumlah-pintu
four       101
two         86
unknown     18
Name: count, dtype: int64

Distribusi body-style:
body-style
sedan          88
hatchback      67
wagon          24
unknown        12
hardtop         8
convertible     6
Name: count, dtype: int64


##3. Mengganti Missing Value pada Variabel Numerik
Untuk variabel numerik yang memiliki data hilang, lakukan **imputasi menggunakan nilai rata-rata (mean)**.

**Variabel numerik pada dataset:**
*   leght
*   width
*   height

**Instruksi:**
*   Hitung nilai rata-rata masing-masing variabel
*   Gunakan nilai tersebut untuk mengganti data yang hilang.




In [75]:
# Mengecek jumlah missing value pada variabel numerik
numerik_cols = ['length', 'width', 'height']
df[numerik_cols].isnull().sum()

,0
length,8
width,5
height,4


In [76]:
# Menghitung nilai rata-rata (mean) dari setiap variabel numerik
mean_length = df['length'].mean()
mean_width = df['width'].mean()
mean_height = df['height'].mean()

# Mengganti missing value pada variabel numerik dengan nilai rata-rata
df['length'] = df['length'].fillna(mean_length)
df['width'] = df['width'].fillna(mean_width)
df['height'] = df['height'].fillna(mean_height)

# Memastikan tidak ada lagi missing value pada variabel numerik
df[numerik_cols].isnull().sum()

,0
length,0
width,0
height,0


##4. Min-Max Normalization
Lakukan **Min-Max Normalization** pada variabel: **wheel-bases**


In [86]:
# Mengecek apakah masih terdapat missing value pada variabel wheel-bases
df['wheel-bases'].isnull().sum()

np.int64(3)

In [89]:
# Mengganti missing value pada variabel wheel-bases dengan nilai rata-rata
df['wheel-bases'] = df['wheel-bases'].fillna(df['wheel-bases'].mean())

In [90]:
# Memastikan tidak ada lagi missing value pada variabel wheel-bases
df['wheel-bases'].isnull().sum()

np.int64(0)

In [91]:
# Melakukan normalisasi Min-Max pada variabel wheel-bases
# agar nilai berada dalam rentang 0 hingga 1
min_val = df['wheel-bases'].min()
max_val = df['wheel-bases'].max()
df['wheel-bases-normalized'] = (df['wheel-bases'] - min_val) / (max_val - min_val)

In [92]:
# Menampilkan perbandingan nilai sebelum dan sesudah normalisasi
df[['wheel-bases', 'wheel-bases-normalized']].head()

,wheel-bases,wheel-bases-normalized
0,88.6,0.058309
1,88.6,0.058309
2,94.5,0.230321
3,99.8,0.384840
4,99.4,0.373178


##5. Z-Score Normalization
Lakukan **Z-score normalization** pada variabel berikut:
*   leght
*   width
*   height


In [93]:
# Melakukan normalisasi Z-score pada variabel numerik
# untuk mengubah distribusi data menjadi memiliki mean = 0 dan standar deviasi = 1
numerik_cols = ['length', 'width', 'height']

for col in numerik_cols:
    mean = df[col].mean()
    std = df[col].std()

    df[col + '_zscore'] = (df[col] - mean) / std

# Menampilkan perbandingan nilai sebelum dan sesudah normalisasi Z-score
df[['length', 'length_zscore',
    'width', 'width_zscore',
    'height', 'height_zscore']].head()

,length,length_zscore,width,width_zscore,height,height_zscore
0,168.8,-0.428526,64.1,-0.867630,48.8,-2.038094
1,168.8,-0.428526,64.1,-0.867630,48.8,-2.038094
2,171.2,-0.226967,65.5,-0.204232,52.4,-0.552035
3,176.6,0.226541,66.2,0.127467,54.3,0.232274
4,176.6,0.226541,66.4,0.222238,54.3,0.232274


##6. Pengolahan Variabel Tambahan
Tahapan ini merupakan preprocessing tambahan di luar instruksi tugas, dilakukan karena variabel **normalized-losses** memiliki jumlah missing value paling tinggi dan merepresentasikan tingkat risiko kendaraan yang penting dalam analisis. Oleh karena itu, dilakukan **imputasi menggunakan mean** agar data tetap lengkap tanpa menghilangkan informasi yang relevan.


In [94]:
# Mengecek apakah masih terdapat missing value pada variabel normalized-losses
df['normalized-losses'].isnull().sum()

np.int64(41)

In [95]:
# Mengganti missing value pada variabel normalized-losses dengan nilai rata-rata
df['normalized-losses'] = df['normalized-losses'].fillna(df['normalized-losses'].mean())

In [96]:
# Memastikan tidak ada lagi missing value pada variabel normalized-losses
df['normalized-losses'].isnull().sum()

np.int64(0)

##7. Dataset Final

In [97]:
# Menampilkan 5 data teratas setelah seluruh proses data cleaning dan normalisasi
df.head()

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height,wheel-bases-normalized,length_zscore,width_zscore,height_zscore
0,122.0,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8,0.058309,-0.428526,-0.867630,-2.038094
1,122.0,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8,0.058309,-0.428526,-0.867630,-2.038094
2,122.0,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4,0.230321,-0.226967,-0.204232,-0.552035
3,164.0,audi,gas,four,sedan,99.8,176.6,66.2,54.3,0.384840,0.226541,0.127467,0.232274
4,164.0,audi,gas,four,sedan,99.4,176.6,66.4,54.3,0.373178,0.226541,0.222238,0.232274


In [99]:
# Menyimpan dataset hasil preprocessing ke file CSV
df.to_csv('data_tugas2_cleaned.csv', index=False)